Add Vertical Advection design document (#301)
E3SM-Project/Omega: components/omega/doc/design/VertAdv.md

The Vertical Advection module is responsible for the vertical transport of mass,
momentum, and tracers within Omega. This implementation is consistent with
Omega's non-Boussinesq governing equations, where the prognostic vertical
coordinate is a pseudo-height, effectively a normalized pressure coordinate. The
methods compute the tendencies associated with the vertical advection terms in
the governing equations, and are designed to allow for flexibility in numerical
scheme configuration.

## 2 Requirements

### 2.1 Requirement: Compute transport pseudo-velocity

Compute the transport pseudo-velocity through vertical layer interfaces based on
the divergence of the horizontal velocity field. Allow for possible movement of
layer interfaces and tilt of layer interfaces

### 2.2 Requirement: Advection scheme flexibility

Support for multiple possible reconstruction schemes, allowing users to select
among different orders of accuracy and upwinding at runtime through
configuration options. Standard and monotonic flux-corrected transport (FCT)
advection algorithms are available for the tendency calculations.

### 2.3 Requirement: Compute tendencies for mass, momentum, and tracers

> Separate methods are neededto compute the vertical-advection tendencies

Reviewer: ```suggestion
Separate methods are needed to compute the vertical-advection tendencies
```

for mass, momentum, and tracers. Mass (`LayerThickness`) and horizontal
momentum (`NormalVelocity`) are stored as 2-D arrays. `LayerThickness` is
cell-centered, while `NormalVelocity` is defined on edges. These tendencies

> use a first-order flux formulation that ensures strict conservation and energy

Reviewer: do we really use first order?  I thought thickness advection was centered (2nd order), but I could be misremembering

Author: You're right, these are centered differences on the interfaces, so they are 2nd order

velocity $d\tilde{w} / d\tilde{z} = - \nabla \cdot {\bf u}.$ In
discrete form, the pseudo-velocity at the top interface of layer $k$ in cell
$i$ is computed from the thickness-weighted divergence of horizontal velocity:

$$
\tilde{w}_{i,k-1/2} = \tilde{w}_{i,k+1/2} +
\frac{1}{A_i} \sum_{e\in EC(i)} n_{e,i} [\tilde{h}_{i,k}]_{e} u_{e,k} l_e,
$$

using the TRiSK formulation of the discrete divergence operator (see
[V0 design document](omega-design-shallow-water-omega0)). The boundary
conditions at the top and bottom interfaces set the pseudo-velocity to zero.

### 3.2 Mass (Thickness)

The time evolution of the pseudo-thickness (represented in the code as
`LayerThickness`) due to vertical transport is

$$
\tilde{h}^{n+1}_{i,k} = \tilde{h}^{n}_{i,k} +
\Delta t \left( \left[\tilde{W}_{tr}\right]_{k+1/2} -
\left[\tilde{W}_{tr}\right]_{k-1/2} \right),
$$

where $\tilde{W}_{tr}$ represents the transport pseudo-velocity and accounts for
the motion of the interface, contributions from the horizontal velocity through
tilted interfaces, and surface source terms.

This first-order form ensures exact conservation of total column thickness and
numerical stability.

> ### 3.3 Momentum (Velocity)

Reviewer: we should call this the velocity equation, it is not momentum as defined

The momentum equation governs the evolution of the horizontal velocity field
(`NormalVelocity`),

$$
u^{n+1}_{e,k} = u^{n}_{e,k} +
\frac{\Delta t}{\left[\tilde{h}_{i,k} \right]_{e}} \left( \left[U \tilde{W}_{tr} \right]_{e,k+1/2} -
\left[U \tilde{W}_{tr} \right]_{e,k-1/2} \right),
$$

> where $U = u - u_k$ removes the component of the horizontal velocity normal to

Reviewer 2: ```suggestion
where $U = u - \tilde{u}$ removes the component of the horizontal velocity normal to
```
Would this be consistent with the v1 gov eqns doc? I see there:

> The second term, $\tilde{u}$, captures the component of horizontal velocity advecting material across a sloping pseudo-height interface

over the active layer interfaces. At each interface, `NormalVelocity` and
`FluxLayerThickEdge` are used to compute `du/dz`, and the `VerticalTransport`
is interpolated from cell centers to edges. The product of these, `w du/dz` is
averaged from interfaces to the middle of each layer and accumulated into the
velocity tendency.

```c++
void VertAdv::computeVelocityVAdvTend(const Array2DReal &Tend,
    const Array2DReal &NormalVelocity, const Array2DReal &FluxLayerThickEdge)
```

The `computeTracerVAdvTend` method serves as an interface that dispatches to
either the standard advection algorithm, or the FCT scheme depending on
runtime configuration.

```c++
void VertAdv::computeTracerVadvTend(const Array3DReal &Tend,
    const Array3DReal &Tracers)
```

The `computeStdVAdvTend` method performs the standard advection update by
looping over the cells and interfaces using nested `parallelForOuter` and
`parallelForInner` constructs. The flux functor chosen from the runtime
configuration is used to compute interface fluxes, which are then applied to
update the tracer tendencies.

The `computeFCTVAdvTend` method implements the flux-corrected transport scheme

> following Zalesak 1979. At each interface, both a high-order flux (chosen via

Reviewer: Are we using zalesak?  isn't that 1st order 2nd order?  I guess the principle is the same though, but maybe cite this too for the high order part?

Reviewer: sorry forgot the citation - https://journals.ametsoc.org/view/journals/mwre/139/9/mwr-d-10-05056.1.xml

Author: Yes, Zalesak is the general concept of FCT, but we use the formulation of Skamarock and Gassmann. Added the citation.

configuration) and a low-order flux (1st-order upwind) are computed. These
fluxes are then blended and limited to ensure monotone transport.

## 5 Verification and Testing

Unit tests are required for each compute method to verify correct operation and

> numerical consistency. In addition, convergence tests should confirm the

Reviewer 2: ```suggestion
numerical consistency. In addition, convergence tests with the [merry-go-round](https://docs.e3sm.org/polaris/main/users_guide/ocean/tasks/merry_go_round.html) test case should confirm the
```

array. The tracer tendencies are computed using a configurable flux-form scheme
that supports multiple interface reconstruction options.

### 2.4 Requirement: Modular design

The vertical advection module must consist of classes and methods that are easily
incorporated into different time-integration schemes and maintain compatibility
with the Omega driver framework.

### 2.5 Desired: Monotonicity checks

The vertical advection module should include optional monotonicity checks to
ensure that reconstructed tracer fields remain physically bounded and free of
spurious oscillations when using the FCT scheme.

## 3 Algorithmic Formulation

The [Omega V1 Governing Equations](omega-design-governing-eqns-omega1) describe
the evolution of mass, velocity, and tracers, including the contributions
from vertical advection.

### 3.1 Pseudo-velocity

The vertical pseudo-velocity is derived from the divergence of the horizontal
velocity $d\tilde{w} / d\tilde{z} = - \nabla \cdot {\bf u}.$ In
discrete form, the pseudo-velocity at the top interface of layer $k$ in cell
$i$ is computed from the thickness-weighted divergence of horizontal velocity:

$$

> \tilde{w}_{i,k-1/2} = \tilde{w}_{i,k+1/2} +

Reviewer 3: Since we are planning on using the "p-star" coordinate, should we explicitly define $$\tilde{W}$$ to include the ALE term?  That would mean adding the pseudo-height/velocity equivalent of equations (4) and (5) from [Petersen et al. 2015](https://doi.org/10.1016/j.ocemod.2014.12.004).

over the edges in the horizontal mesh, and a `parallelForInner` loop iterates
over the active layer interfaces. At each interface, `NormalVelocity` and
`FluxLayerThickEdge` are used to compute `du/dz`, and the `VerticalTransport`
is interpolated from cell centers to edges. The product of these, `w du/dz` is
averaged from interfaces to the middle of each layer and accumulated into the
velocity tendency.

```c++
void VertAdv::computeVelocityVAdvTend(const Array2DReal &Tend,
    const Array2DReal &NormalVelocity, const Array2DReal &FluxLayerThickEdge)
```

The `computeTracerVAdvTend` method serves as an interface that dispatches to
either the standard advection algorithm, or the FCT scheme depending on
runtime configuration.

```c++
void VertAdv::computeTracerVadvTend(const Array3DReal &Tend,
    const Array3DReal &Tracers)
```

The `computeStdVAdvTend` method performs the standard advection update by
looping over the cells and interfaces using nested `parallelForOuter` and
`parallelForInner` constructs. The flux functor chosen from the runtime
configuration is used to compute interface fluxes, which are then applied to
update the tracer tendencies.

> The `computeFCTVAdvTend` method implements the flux-corrected transport scheme

Reviewer 3: Maybe make it more clear here and in Section 3.4 that FCT is only for tracers?
